src/core/lib/iomgr/polling_interface.h

```diff
@@ -0,0 +1,167 @@
+/*
+ *
+ * Copyright 2015 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef GRPC_CORE_LIB_IOMGR_POLLING_INTERFACE_H
+#define GRPC_CORE_LIB_IOMGR_POLLING_INTERFACE_H
+
+#include <grpc/support/port_platform.h>
+#include <grpc/support/sync.h>
+#include <grpc/support/time.h>
+
+#include "src/core/lib/iomgr/exec_ctx.h"
+
+#ifndef NDEBUG
+extern grpc_tracer_flag grpc_trace_fd_refcount;
+#endif
+
+namespace grpc_core {
+
+// Forward declarations
+class PollingJoin;
+
+// A Pollable is an object that can be polled
+// Polling engines derive from this interface and implement
+// This interface should not be implemented outside of a polling engine
+class Pollable {};
+
+// A PollableCollection is a front-end interface to Poller & PollingJoin,
+// collecting common methods.
+// In many situations we don't care which we're dealing with, and so this class
+// serves to abstract away that detail
+class PollableCollection {
+ public:
+  // Add a Pollable to this collection
+  virtual void AddPollable(Pollable* pollable) = 0;
+
+ protected:
+  PollableCollection() {}
+  ~PollableCollection() {}
+
+ private:
+  friend class PollingJoin;
+  virtual void AddToPollingJoin(PollingJoin* join) = 0;
+};
+
+// A Poller is a set of file descriptors that a higher level item is
+// interested in, and a method to poll on them. For example:
+//    - a server will typically keep a poller containing all connected channels,
+//      so that it can find new calls to service
+//    - a completion queue might keep a poller with an entry for each transport
+//      that is servicing a call that it's tracking
+class Poller : public PollableCollection {
```

Comment:
I feel the name `Pollset` is better left unchanged. The name `Poller` to me implies that is 'polling' something (and makes me mentally associate 'Poller' to a thread); The name `Pollset` seems to imply that this is a "set" of something.

Reply:
I've had (so many) discussions in the past where folks were surprised that a pollset took an active role in polling, which is what I'm trying to address here. 

This type does take an active (if not continuous) part in the polling process, and I'd like to highlight it in the name - so much so that I'm thinking about changing Work() to Poll() to further highlight the fact.

Reply:
I for one would welcome s/Work/Poll.

Reply:
Changed Poll --> Work, which leaves the current open question being is Poller the right name for this?

Given the switch to PollableSet, I'm hesitant to call this Pollset or PollingSet